### ISSUE TITLE
Not much information on tensor_crop

### ISSUE THREAD

**Issue by AnanthaRaman123** [nnstreamer/nnstreamer#3750]
# There isn't much information or clear examples on how to use tensor_crop element
I am trying to find an answer for a problem statement as follows. I want to basically crop out the bounding-box region created by tensor_filter / tensor_decoder . I thought to use the nnstreamer tensor_crop for the same. But its getting hard to exactly understand how to use it as there isn't much information or clear cut examples on how to use the same.

Expected Result
============
Few tensor_crop related examples (Either bash scripts/python scrips)

How to Reproduce
===============
Nothing

**Comment by myungjoo** [comment]
@jaeyun-jung If we arleady have an example code, please link it (here and at /gst/nstreamer/tensor_crop/*.c 's main boilerplate)

And you may provide more examples in the LINE 14-36 area as well.

**Comment by AnanthaRaman123** [comment]
Any expected date to get the answer ??

**Comment by jaeyun-jung** [comment]
We don't have a pipeline example using tensor_crop yet.
(considering to add new tensor-decoder subplugin for object detection results)

Now you can just find simple testcases for tensor_crop,
https://github.com/nnstreamer/nnstreamer/blob/2a42dda5b773d4c1a7ba29db7faa611f6eb0c6fe/tests/nnstreamer_plugins/unittest_plugins.cc#L7171


tensor_crop has 2 sink-pads and 1 src-pad, and this handles flexible tensor.
(About flexible-tensor, see https://nnstreamer.github.io/data-type-and-flow-control.html#othertensorsformatflexible)

1. src-pad
[src] output is a flexible-tensor with N memories of cropped data.

2. sink-pad
[raw] input buffer is a tensor.
[info] crop-info, input buffer pushed into 'info' pad should be a flexible-tensor.

For example, 
Input buffer :
- raw buffer : uint32 [1, 2, ..., 40] with dimension 1:10:4:1
- info buffer : 2 set of crop-info (x,y,w,h), uint32 [3, 0, 3, 1, 2, 1, 7, 2]

Output buffer has 2 memory blocks, each memory has (header + data).
1st flexible tensor : header + result [4, 5, 6]
2nd flexible tensor : header + result [13, 14, ..., 19, 23, 24, ..., 29]

**Comment by AnanthaRaman123** [comment]
Its kind of actually not so easy to create a working pipeline example from the TestCase. I tried few methods, none of them which worked and thats where I posted this question. Is it not possible to create a sample example, say an example where object is detected and that object is cropped using tensor_crop. Is it possible for you to create that ?

**Comment by jaeyun-jung** [comment]
No, impossible now. The testcase I shared is just to explain the data structure of crop-info and how crop-info would be generated (flexible tensor).

**Comment by jaeyun-jung** [comment]
Real example pipeline will be complex, and require new decoder sub-plugins.
e.g., 
1. implement new plugin : decoder_object_detection
 - this generates crop-info from inference results.
2. muxing raw video tensor and crop-info tensor.
3. demuxing cropped tensor + generating video frame for each cropped data
...


tensor_mux ! tensor_crop ! ...
[CAM] ! videoconvert ! videoscale ! tensor_converter ! tee name=t
t. ! queue ! mux.sink_0 
t. ! queue ! tensor_filter model=objectdetection_model ! tensor_decoder mode=object_detection ! mux.sink_1

**Comment by jaeyun-jung** [comment]
What I want to say is, without a specific plugin (or custom filter?), it is difficult to meetup all usecase. tensor-crop is just an element to get the cropped data - it is clearly unfixed data size, so uses flexible tensor now.

**Comment by myungjoo** [comment]
@jaeyun-jung After 7.0 M2 release (and RTOS builds of nns-edge), could you please let someone (or yourself) write a working pipeline example using crop (w/o appsink) to do

crop a found object (human? :) )  and imagesink the found object rescaled to a static size (e.g., 640x480).

**Comment by bruecks** [comment]
The output of the crop will be other/tensors, so it will be a collection of at most 16 tensors. This seems to me a small number because there might be hundreds of faces in the picture. Will the requirement that the maximum number of tensors is 16 be dropped?

The number of tensors is unknown because the number of detected faces will change from frame to frame. Does the tensor demux support a variable number of tensors? This means that the number of src from the demux pad would be variable. Is it possible to specify a gstreamer pipeline if the number of sources is variable?

**Comment by myungjoo** [comment]
> The output of the crop will be other/tensors, so it will be a collection of at most 16 tensors. This seems to me a small number because there might be hundreds of faces in the picture. Will the requirement that the maximum number of tensors is 16 be dropped?
> 
> The number of tensors is unknown because the number of detected faces will change from frame to frame. Does the tensor demux support a variable number of tensors? This means that the number of src from the demux pad would be variable. Is it possible to specify a gstreamer pipeline if the number of sources is variable?

Each tensor is assigned to each GStreamer Buffer, whose limit is 16 and can be changed by updating a header of GStreamer.
We've chosen to use a GstBuf for each tensor to minimize memcpy during stream path manipulations (mux, demux, ...)

However, if it is the case of cropping, you can place cropped images into a single tensor. For example, if you have N cropped images of Width:Height:CH, you can have one tensor of N:Width(max):Height(max):CH, not N tensors of Width:Height:CH. Unless you want to separate streams paths of each tensor, the former results in simpler (and higher performing) pipelines. To process N times of Width:Height:CH tensors afterwards, you can slap in tensor_aggrgate to create N frames of W:H:CH tensors. The concern here is that N needs to be static here, where you may slap in tensor_if to remove "noop" frames from the stream before pushing them into  another tensor_filter.

**Comment by bruecks** [comment]
It is actually a concern that N must be static if I imagine the case that, say usually we have only one or two faces in a picture but in rare cases it might be, say, one hundred. We would always have to allocate full memory. My idea was to process the N images one by one in a special loop in the tensor_filter. Unfortunately I do not see a way to tell the invoke function what the size of my dynamic input tensor is because it is not passed to that function and inside the invoke, it assumes the size is the same as the network requires, as far as I see it in the tvm filter.

**Comment by bruecks** [comment]
I asked in a separate issue if it is possible to have a flexible tensor output where the dimensions change from frame to frame and how that can be implemented in the invoke function of the tensor filter. https://github.com/nnstreamer/nnstreamer/issues/3926

**Comment by myungjoo** [comment]
Closing: the requested working example now exists. The tensor_region decoder subplugin (ext/nnstreamer/tensor_decoder/tensordec-tensor_region.c, added in 2023) generates crop-info for tensor_crop, and tests/nnstreamer_decoder_tensor_region/runTest.sh contains a full object-detection to tensor_region to tensor_crop pipeline. gsttensor_crop.c also carries an example launch line in its documentation.
